src/runner.py

```diff
@@ -85,599 +56,438 @@ class LxdInstanceConfig(TypedDict):
+            )
 
-class RunnerRemoveError(RunnerError):
-    """Error for runner removal failure."""
+        if self.instance is None:
+            return
 
+        if self.instance.status == "Running":
+            try:
+                self.instance.stop(wait=True, timeout=60)
+            except LXDAPIException:
+                logger.exception("Unable to gracefully stop runner within timeout.")
+                self.instance.stop(force=True)
 
-class RunnerStartError(RunnerError):
-    """Error for runner start failure."""
+            with suppress(LXDAPIException):
+                # Ephemeral containers should auto-delete when stopped;
+                # this is just a fall-back.
+                self.instance.delete(wait=True)
+        else:
+            # We somehow have a non-running instance which should have been
+            # ephemeral. Try to delete it and allow any errors doing so to
+            # surface.
+            try:
+                self.instance.delete(wait=True)
+            except LXDAPIException as err:
+                raise RunnerRemoveError(f"Unable to remove {self.config.name}") from err
 
+    @retry(tries=5, delay=1, local_logger=logger)
+    def _create_instance(
+        self, image: str, resources: VirtualMachineResources, ephemeral: bool = True
+    ) -> pylxd.models.Instance:
+        """Create an instance of runner.
 
-class RunnerManager:
-    """Class for querying and controlling the runners.
+        Args:
+            image: Image used to launch the instance hosting the runner.
+            resources: Configuration of the virtual machine resources.
+            ephemeral: Whether the instance is ephemeral.
 
-    Attributes:
-        proxies (Dict[str, str]): Mapping of proxy env vars.
-        path (str): GitHub repository path in the format '<org>/<repo>', or the GitHub org name.
-        app_name (str): An name for the set of runners.
-        reconcile_interval (int): Number of minutes between each reconciliation of runners state.
-    """
+        Returns:
+            LXD instance of the runner.
+        """
+        logger.info("Creating an LXD instance for runner: %s", self.config.name)
 
-    # TODO: Verify if this violates bandit B108.
-    runner_bin_path = Path("/var/cache/github-runner-operator/runner.tgz")
-    env_file = Path("/opt/github-runner/.env")
+        self._ensure_runner_profile()
+        resource_profile = self._get_resource_profile(resources)
 
-    def __init__(self, path: str, token: str, app_name: str, reconcile_interval: int) -> None:
-        """Construct RunnerManager object for creating and managing runners.
+        # Create runner instance.
+        instance_config: LxdInstanceConfig = {
+            "name": self.config.name,
+            "type": "virtual-machine",
+            "source": {
+                "type": "image",
+                "mode": "pull",
+                "server": "https://cloud-images.ubuntu.com/daily",
```

Comment:
Should we use a more stable version of Ubuntu instead of daily.

Reply:
Good idea.
Is there a link for the current stable release, or should I change this to say `jammy`?

Reply:
There isn't a link for stable release, so jammy is a good idea.